merkle/smt/tiles.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smt
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/storage/tree"
+)
+
+// TileSet represenst a set of Merkle tree tiles and the corresponding nodes.
+//
+// TODO(pavelkalinnikov): Make it immutable.
+type TileSet struct {
+	layout *tree.Layout
+	tiles  map[tree.NodeID2][]Node
+	hashes map[tree.NodeID2][]byte
+	h      mapHasher
+}
+
+// NewTileSet creates an empty TileSet with the given tree parameters.
+func NewTileSet(treeID int64, hasher hashers.MapHasher, layout *tree.Layout) *TileSet {
+	tiles := make(map[tree.NodeID2][]Node)
+	hashes := make(map[tree.NodeID2][]byte)
+	h := bindHasher(hasher, treeID)
+	return &TileSet{layout: layout, tiles: tiles, hashes: hashes, h: h}
+}
+
+// Hashes returns a map containing all node hashes keyed by node IDs.
+func (t *TileSet) Hashes() map[tree.NodeID2][]byte {
+	return t.hashes
+}
+
+// Add puts the given tile into the set.
+//
+// TODO(pavelkalinnikov): Take a whole list of Tiles instead.
```

Comment:
Should probably comment that it must be used single-threaded (or pop a `Mutex` in to lock the `map`s)